types/src/crypto/asymmetric_key.rs

```diff
@@ -133,15 +133,19 @@ impl SecretKey {
     pub const SECP256K1_LENGTH: usize = SECP256K1_SECRET_KEY_LENGTH;
 
     /// Constructs a new Ed25519 variant from a byte array.
-    pub fn ed25519(bytes: [u8; Self::ED25519_LENGTH]) -> Self {
-        // safe to unwrap as `SecretKey::from_bytes` can only fail if the provided slice is the
-        // wrong length.
-        SecretKey::Ed25519(ed25519_dalek::SecretKey::from_bytes(&bytes).unwrap())
+    pub fn ed25519(bytes: [u8; Self::ED25519_LENGTH]) -> Result<Self, Error> {
+        // while it is safe to unwrap as `SecretKey::from_bytes` can only fail if the provided slice
+        // is the wrong length.
+        // We should be disciplined about unwrapping in constructors.
+        match ed25519_dalek::SecretKey::from_bytes(&bytes) {
+            Ok(key) => Ok(SecretKey::Ed25519(key)),
+            Err(error) => Err(Error::AsymmetricKey(error.to_string())),
```

Comment:
Can we make a `From` impl for these `ed25519_dalek` errors?  We should just have variants that wrap them if possible rather than have these strings.

Reply:
I think our unusual approach of just making the errors a string in this case is justified.  The [`ed25519::Error` type](https://docs.rs/signature/1.2.2/signature/struct.Error.html) and [`k256::Error` type](https://docs.rs/elliptic-curve/0.9.6/src/elliptic_curve/error.rs.html#10) are deliberately opaque, and the resulting log messages extremely unhelpful.

I don't think we should just do `Err(Error::AsymmetricKey(error.to_string()))` though, as that will just produce the unhelpful text "signature error".  Instead I think we should have something like "failed to construct Ed25519 secret key from bytes" which leaks no details about the secret key bytes, but is specific enough to be meaningful for debugging purposes.